File: backend-v5/app/schemas/notifications.py
"""Pydantic schemas for notification endpoints."""
from datetime import datetime, time
from typing import Literal, Optional

from pydantic import BaseModel, Field


NotificationSeverity = Literal["low", "medium", "high", "critical"]
NotificationReasonCode = Literal[
    "major_nepal",
    "followed_topic",
    "followed_district",
    "followed_province",
    "home_district",
]
NotificationTab = Literal["all", "for_you", "major"]


class NotificationResponse(BaseModel):
    """Single notification."""

    id: str
    type: str
    category: str
    severity: Optional[str] = None
    title: str
    message: Optional[str] = None
    data: Optional[dict] = None
    reason_code: Optional[str] = None
    reason_label: Optional[str] = None
    topic_id: Optional[str] = None
    province_name: Optional[str] = None
    district_name: Optional[str] = None
    source_kind: Optional[str] = None
    source_id: Optional[str] = None
    deeplink_url: Optional[str] = None
    is_read: bool
    created_at: datetime


class NotificationListResponse(BaseModel):
    """List of notifications with unread count."""

    items: list[NotificationResponse]
    unread_count: int = 0


class NotificationFeedResponse(NotificationListResponse):
    """Cursor-paginated notification feed response."""

    next_cursor: Optional[str] = None
    tab: NotificationTab = "all"


class NotificationPreferencesResponse(BaseModel):
    """Server-side notification preferences."""

    notifications_enabled: bool = True
    include_major_alerts: bool = True
    min_severity: NotificationSeverity = "high"
    home_district: Optional[str] = None
    followed_districts: list[str] = []
    followed_provinces: list[str] = []
    followed_topics: list[str] = []
    muted_districts: list[str] = []
    muted_provinces: list[str] = []
    muted_topics: list[str] = []
    quiet_hours_start: Optional[time] = None
    quiet_hours_end: Optional[time] = None
    has_saved_preferences: bool = False


class NotificationPreferencesUpdate(BaseModel):
    """Update request for notification preferences."""

    notifications_enabled: Optional[bool] = None
    include_major_alerts: Optional[bool] = None
    min_severity: Optional[NotificationSeverity] = None
    home_district: Optional[str] = None
    followed_districts: Optional[list[str]] = None
    followed_provinces: Optional[list[str]] = None
    followed_topics: Optional[list[str]] = None


class NotificationFollowSimilarResponse(BaseModel):
    """Result of follow-similar or mute actions."""

    success: bool = True
    preferences: NotificationPreferencesResponse


class MarkReadResponse(BaseModel):
    """Response after marking notification(s) as read."""

    success: bool = True
    marked_read: int = 0


class NotificationSeedRequest(BaseModel):
    """Dev-only helper payload for seeding test notifications."""

    count: int = Field(default=3, ge=1, le=10)

File: frontend/src/api/notifications.ts
import { apiClient } from './client'

export type NotificationSeverity = 'low' | 'medium' | 'high' | 'critical'
export type NotificationReasonCode =
  | 'major_nepal'
  | 'followed_topic'
  | 'followed_district'
  | 'followed_province'
  | 'home_district'
export type NotificationTab = 'all' | 'for_you' | 'major'

export interface Notification {
  id: string
  type: string
  category: string
  severity: NotificationSeverity | null
  title: string
  message: string | null
  data: Record<string, any> | null
  reason_code: NotificationReasonCode | null
  reason_label: string | null
  topic_id: string | null
  province_name: string | null
  district_name: string | null
  source_kind: string | null
  source_id: string | null
  deeplink_url: string | null
  is_read: boolean
  created_at: string
}

export interface NotificationListResponse {
  items: Notification[]
  unread_count: number
}

export interface NotificationFeedResponse extends NotificationListResponse {
  next_cursor: string | null
  tab: NotificationTab
}

export interface NotificationPreferences {
  notifications_enabled: boolean
  include_major_alerts: boolean
  min_severity: NotificationSeverity
  home_district: string | null
  followed_districts: string[]
  followed_provinces: string[]
  followed_topics: string[]
  muted_districts: string[]
  muted_provinces: string[]
  muted_topics: string[]
  quiet_hours_start: string | null
  quiet_hours_end: string | null
  has_saved_preferences: boolean
}

export interface NotificationPreferencesUpdate {
  notifications_enabled?: boolean
  include_major_alerts?: boolean
  min_severity?: NotificationSeverity
  home_district?: string | null
  followed_districts?: string[]
  followed_provinces?: string[]
  followed_topics?: string[]
}

export async function fetchNotifications(): Promise<NotificationListResponse> {
  const { data } = await apiClient.get('/notifications/')
  return data
}

export async function fetchNotificationFeed(params: {
  tab?: NotificationTab
  unreadOnly?: boolean
  limit?: number
  cursor?: string | null
}): Promise<NotificationFeedResponse> {
  const { data } = await apiClient.get('/notifications/feed', {
    params: {
      tab: params.tab ?? 'all',
      unread_only: params.unreadOnly ?? false,
      limit: params.limit ?? 25,
      cursor: params.cursor ?? undefined,
    },
  })
  return data
}

export async function fetchNotificationPreferences(): Promise<NotificationPreferences> {
  const { data } = await apiClient.get('/notifications/preferences')
  return data
}

export async function updateNotificationPreferences(
  payload: NotificationPreferencesUpdate,
): Promise<NotificationPreferences> {
  const { data } = await apiClient.put('/notifications/preferences', payload)
  return data
}

export async function markNotificationRead(notificationId: string): Promise<{ success: boolean }> {
  const { data } = await apiClient.post(`/notifications/${notificationId}/read`)
  return data
}

export async function markAllNotificationsRead(): Promise<{ success: boolean; marked_read: number }> {
  const { data } = await apiClient.post('/notifications/read-all')
  return data
}

export async function muteNotification(notificationId: string): Promise<{ success: boolean; preferences: NotificationPreferences }> {
  const { data } = await apiClient.post(`/notifications/${notificationId}/mute`)
  return data
}

export async function followSimilarNotification(notificationId: string): Promise<{ success: boolean; preferences: NotificationPreferences }> {
  const { data } = await apiClient.post(`/notifications/${notificationId}/follow-similar`)
  return data
}
